akka-stream/src/main/scala/akka/stream/stage/Stage.scala

```diff
@@ -91,6 +91,15 @@ abstract class AbstractStage[-In, Out, PushD <: Directive, PullD <: Directive, C
   }
 
   /**
+   * User overridable callback.
+   * <p/>
+   * It is called before any other method defined on the `Stage`.
+   * Empty default implementation.
+   */
+  @throws(classOf[Exception])
+  def preStart(ctx: LifeCtx): Unit = ()
```

Comment:
This allows subtypes to refine the preStart context independently of the normal context. So users don't get "push" here, it would not make any sense.